### PR TITLE
GEOMETRY-119: adding Vector.normalizeOrNull() method

### PR DESCRIPTION
**Pull request by darkma773r**

### CODE REVIEW

File: commons-geometry-core/src/main/java/org/apache/commons/geometry/core/Vector.java [hunk 1]
```diff
@@ -93,11 +93,22 @@
 
     /** Get a normalized vector aligned with the instance. The returned
      * vector has a magnitude of 1.
-     * @return a new normalized vector
-     * @throws IllegalArgumentException if the norm is zero, NaN, or infinite
+     * @return normalized vector
+     * @throws IllegalArgumentException if the norm (or norm inverse, if applicable)
+     *      is zero, NaN, or infinite
```
Comment (arturobernalg) [inline commons-geometry-core/src/main/java/org/apache/commons/geometry/core/Vector.java:98]: This space is correct?

**Review by arturobernalg — APPROVED** [review]
LGTM

File: commons-geometry-core/src/main/java/org/apache/commons/geometry/core/Vector.java (hunk 1 above)
Comment (darkma773r) [reply commons-geometry-core/src/main/java/org/apache/commons/geometry/core/Vector.java:98]: I find it easier to read when the second line of tag content is indented. It doesn't affect the generated docs so it's a personal preference thing. Is that what you're referring to?

**Comment by singhbaljit** [comment]
IMO, it shouldn't return `null`. Instead, it should return `Optional<V>`.

**Comment by darkma773r** [comment]
@singhbaljit, that's a good idea. Can you suggest that on the JIRA issue along with what you think we should name the method? Would it still be "normalizeOrNull"?

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Vector3D.java [hunk 2]
```diff
@@ -760,6 +766,21 @@ public static Vector3D linearCombination(final double a1, final Vector3D v1,
...
 
+        /** Maximum coordinate value for computing normalized vectors
+         * with raw, unscaled values.
+         */
+        private static final double UNSCALED_MAX = 0x1.0p+500;
+
+        /** Factor used to scale up coordinate values in order to produce
+         * normalized coordinates without overflow or underflow.
+         */
+        private static final double SCALE_UP_FACTOR = 0x1.0p600;
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Vector3D.java:777]: For consistency use a '+' here before the 600 or drop it from the UNSCALED_MAX constant.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Vector3D.java [hunk 3]
```diff
@@ -828,5 +826,88 @@ public Vector3D withNorm(final double mag) {
...
+         *      is false
+         * @throws IllegalArgumentException if the computed norm is zero, NaN, or infinite
+         */
+        private static Unit tryCreateNormalized(final double x, final double y, final double z,
+                final boolean throwOnFailure) {
+
+            // Compute the inverse norm directly. If the result is a non-zero real number,
+            // then we can go ahead and construct the unit vector immediately. If not,
+            // we'll do some extra work for edge cases.
+            final double norm = Math.sqrt((x * x) + (y * y) + (z * z));
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Vector3D.java:871]: Handling the common case first is a better solution than checking for scaling immediately.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/Vector2D.java [hunk 4]
```diff
@@ -686,6 +692,21 @@ public static Vector2D linearCombination(final double a1, final Vector2D v1,
...
 
+        /** Maximum coordinate value for computing normalized vectors
+         * with raw, unscaled values.
+         */
+        private static final double UNSCALED_MAX = 0x1.0p+500;
+
+        /** Factor used to scale up coordinate values in order to produce
+         * normalized coordinates without overflow or underflow.
+         */
+        private static final double SCALE_UP_FACTOR = 0x1.0p600;
```
Comment (aherbert) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/Vector2D.java:703]: Update constants as noted before

File: commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/threed/Vector3DTest.java [hunk 5]
```diff
@@ -396,16 +395,40 @@ public void testNormalize() {
...
+        checkVector(Vector3D.of(0, 0, Double.MIN_VALUE).normalize(), 0, 0, 1);
+
+        checkVector(Vector3D.of(Double.MIN_VALUE, Double.MIN_VALUE, Double.MIN_VALUE).normalize(),
+                invSqrt3, invSqrt3, invSqrt3);
+
+        checkVector(Vector3D.of(Double.MIN_NORMAL, 0, 0).normalize(), 1, 0, 0);
+        checkVector(Vector3D.of(0, Double.MIN_NORMAL, 0).normalize(), 0, 1, 0);
+        checkVector(Vector3D.of(0, 0, Double.MIN_NORMAL).normalize(), 0, 0, 1);
+
+        checkVector(Vector3D.of(Double.MIN_NORMAL, Double.MIN_NORMAL, Double.MIN_NORMAL).normalize(),
```
Comment (aherbert) [inline commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/threed/Vector3DTest.java:409]: Use a negative here for one of the values (as done using MAX_VALUE below).

File: commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/twod/Vector2DTest.java [hunk 6]
```diff
@@ -317,21 +317,43 @@ public void testSubtract_scaled() {
...
+
+        checkVector(Vector2D.of(Double.MIN_VALUE, 0).normalize(), 1, 0);
+        checkVector(Vector2D.of(0, Double.MIN_VALUE).normalize(), 0, 1);
+
+        checkVector(Vector2D.of(Double.MIN_VALUE, Double.MIN_VALUE).normalize(), invSqrt2, invSqrt2);
+
+        checkVector(Vector2D.of(Double.MIN_NORMAL, 0).normalize(), 1, 0, 0);
+        checkVector(Vector2D.of(0, Double.MIN_NORMAL).normalize(), 0, 1, 0);
+
+        checkVector(Vector2D.of(Double.MIN_NORMAL, Double.MIN_NORMAL).normalize(), invSqrt2, invSqrt2);
```
Comment (aherbert) [inline commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/twod/Vector2DTest.java:338]: Use a negative for one of the values.

File: commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/twod/Vector2DTest.java [hunk 7]
```diff
@@ -345,6 +367,40 @@ public void testNormalize_isIdempotent() {
...
+        checkVector(Vector2D.of(100, 0).normalizeOrNull(), 1, 0);
+        checkVector(Vector2D.of(-100, 0).normalizeOrNull(), -1, 0);
+
+        checkVector(Vector2D.of(2, 2).normalizeOrNull(), invSqrt2, invSqrt2);
+        checkVector(Vector2D.of(-2, -2).normalizeOrNull(), -invSqrt2, -invSqrt2);
+
+        checkVector(Vector2D.of(Double.MIN_VALUE, 0).normalizeOrNull(), 1, 0);
+        checkVector(Vector2D.of(0, Double.MIN_VALUE).normalizeOrNull(), 0, 1);
+
+        checkVector(Vector2D.of(Double.MAX_VALUE, -Double.MAX_VALUE).normalizeOrNull(), invSqrt2, -invSqrt2);
```
Comment (aherbert) [inline commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/twod/Vector2DTest.java:385]: There is no test using two MIN_NORMAL coordinates as is done in the testNormalise test

File: commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/threed/Vector3DTest.java [hunk 8]
```diff
@@ -419,6 +442,42 @@ public void testNormalize_isIdempotent() {
...
+        checkVector(Vector3D.of(-100, 0, 0).normalizeOrNull(), -1, 0, 0);
+
+        checkVector(Vector3D.of(2, 2, 2).normalizeOrNull(), invSqrt3, invSqrt3, invSqrt3);
+        checkVector(Vector3D.of(-2, -2, -2).normalizeOrNull(), -invSqrt3, -invSqrt3, -invSqrt3);
+
+        checkVector(Vector3D.of(Double.MIN_VALUE, 0, 0).normalizeOrNull(), 1, 0, 0);
+        checkVector(Vector3D.of(0, Double.MIN_VALUE, 0).normalizeOrNull(), 0, 1, 0);
+        checkVector(Vector3D.of(0, 0, Double.MIN_VALUE).normalizeOrNull(), 0, 0, 1);
+
+        checkVector(Vector3D.of(-Double.MAX_VALUE, -Double.MAX_VALUE, -Double.MAX_VALUE).normalizeOrNull(),
```
Comment (aherbert) [inline commons-geometry-euclidean/src/test/java/org/apache/commons/geometry/euclidean/threed/Vector3DTest.java:461]: There is no test using three MIN_NORMAL coordinates as is done in the testNormalise test

**Review by aherbert — COMMENTED** [review]
Hi @darkma773r,  just a few nitpicks to address here.

**Comment by darkma773r** [comment]
Updated. Thanks, @aherbert!

**Pull request merged**